indexer/SymbolFormatter.cc

```diff
@@ -40,44 +41,197 @@ std::string_view SymbolFormatter::getMacroSymbol(clang::SourceLocation defLoc) {
+  // Pending:
+  // - OMPDeclareReductionDecl
+  // - OMPDeclareMapperDecl
+  // - FunctionDecl
+  // - ObjCMethodDecl
+  // - ObjCContainerDecl
+  // - LinkageSpecDecl
+  // - ExportDecl
+  // - BlockDecl
+  // - CapturedDecl
+  return std::nullopt;
+}
+
+std::optional<std::string_view>
+SymbolFormatter::getTagSymbol(const clang::TagDecl *tagDecl) {
+  return this->getSymbolCached(tagDecl, [&]() -> std::string {
+    auto optContextSymbol = this->getContextSymbol(tagDecl->getDeclContext());
+    if (!optContextSymbol.has_value()) {
+      return {};
+    }
+    auto contextSymbol = optContextSymbol.value();
+    if (!tagDecl->getDeclName().isEmpty()) {
+      llvm::raw_string_ostream os(this->scratchBuffer);
+      static_cast<const clang::NamedDecl *>(tagDecl)->printName(os);
+      std::string out{fmt::format("{}{}#", contextSymbol, this->scratchBuffer)};
+      this->scratchBuffer.clear();
+      return out;
+    }
+    auto definitionTagDecl = tagDecl->getDefinition();
+    ENFORCE(definitionTagDecl, "can't forward-declare an anonymous type");
+    auto defLoc =
+        this->sourceManager.getExpansionLoc(definitionTagDecl->getLocation());
+
+    auto defFileId = this->sourceManager.getFileID(defLoc);
+    ENFORCE(defFileId.isValid());
+    auto counter = this->anonymousTypeCounters[{defFileId}]++;
+
+    auto declContext = definitionTagDecl->getDeclContext();
+    if (llvm::isa<clang::NamespaceDecl>(declContext)
+        || llvm::isa<clang::TranslationUnitDecl>(declContext)) {
+      // If the anonymous type is inside a namespace, then we know the
+      // DeclContext chain only has namespaces (types cannot contain
+      // namespaces). So include the file hash too, to avoid collisions across
+      // files putting anonymous types into the same namespace. For example,
+      //
+      //   // A.h
+      //   namespace z { struct { void f() {} } x; }
+      //   // B.h
+      //   namespace z { struct { int f() { return 0; } } y; }
+      //
+      // If we don't include the hash, the anonymous structs will end up with
+      // the same symbol name.
+      if (auto optRelativePath = this->getCanonicalPath(defFileId)) {
+        HashValue hashValue{0};
+        auto sv = optRelativePath->asStringView();
+        hashValue.mix(reinterpret_cast<const uint8_t *>(sv.data()), sv.size());
+        return fmt::format("{}$anontype_{:x}_{}#", contextSymbol,
+                           hashValue.rawValue, counter);
+      }
+    }
+    return fmt::format("{}$anontype_{}#", contextSymbol, counter);
+  });
+}
+
+std::optional<std::string_view> SymbolFormatter::getEnumConstantSymbol(
+    const clang::EnumConstantDecl *enumConstantDecl) {
+  return this->getSymbolCached(enumConstantDecl, [&]() -> std::string {
+    auto parentEnumDecl =
+        llvm::dyn_cast<clang::EnumDecl>(enumConstantDecl->getDeclContext());
+    ENFORCE(parentEnumDecl,
+            "decl context for EnumConstantDecl should be EnumDecl");
+    if (!parentEnumDecl) {
```

Comment:
Since you have ENFORCEd this, can you simply crash by now and not worry about the early return?

Reply:
The ENFORCE will not run in a release build, but I don't want it to crash in a release build, because this is not a serious error, we can skip whatever higher-level operation was being performed (e.g. emitting a def/ref).

Reply:
When will it happen?

Any plan for counting these? When this is all done and the feedback is "make it better" having stats on cases like this would be very valuable and a lot of toil to re-discover.

Reply:
There is some infrastructure in Sorbet/scip-ruby for more observability, like timing for ENFORCEs (so that maybe we can turn some of them to be always on), and counters like the one you mentioned. I have not finished porting all of that here yet. My thinking was to resume that sometime after the MVP is done (or around it). We'd also need to figure out how to collect/analyze the extra data being emitted.